### PR TITLE
Improvements for sp template support

### PR DESCRIPTION
**Pull request by IndunilRathnayake**
### Proposed changes in this pull request
Code improvements for sp template support feature.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/add-service-provider-as-template.jsp [hunk 1]
```diff
@@ -17,20 +17,36 @@
...
-<%@ page import="org.wso2.carbon.identity.application.mgt.ui.util.ApplicationMgtUIUtil" %>
-<%@ page import="org.wso2.carbon.ui.CarbonUIUtil" %>
-<%@ page import="org.wso2.carbon.utils.ServerConstants" %>
 <%@ page import="org.wso2.carbon.identity.application.mgt.ui.client.ApplicationManagementServiceClient" %>
-<%@ page import="org.wso2.carbon.identity.application.common.model.xsd.SpTemplate" %>
+<%@ page import="org.wso2.carbon.identity.application.mgt.ui.util.ApplicationMgtUIUtil" %>
 <%@ page import="org.wso2.carbon.ui.CarbonUIMessage" %>
+<%@ page import="org.wso2.carbon.ui.CarbonUIUtil" %>
+<%@ page
+        import="org.wso2.carbon.utils.ServerConstants" %>
```
Comment (thanujalk) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/add-service-provider-as-template.jsp]: format this line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/add-sp-template-finish-ajaxprocessor.jsp [hunk 2]
```diff
@@ -47,24 +61,35 @@
...
-<script>
-    location.href = 'add-sp-template.jsp';
-</script>
+    <script>
+        location.href = 'add-sp-template.jsp';
+    </script>
 <%
     }
-%>
+%>
```
Comment (thanujalk) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/add-sp-template-finish-ajaxprocessor.jsp]: add new line at the end

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 3]
```diff
@@ -387,12 +393,86 @@
...
+            if (!isValied) {
+                return false;
+            }
+        }
+        if ($('.roleMapIdp').length > 0) {
+            var isValied = true;
+            $.each($('.roleMapIdp'), function () {
+                if ($(this).val().length == 0) {
+                    isValied = false;
+                    CARBON.showWarningDialog('Please complete Role Mapping Configuration section');
```
Comment (thanujalk) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: can we get this messages from a resource bundle?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/client/ApplicationManagementServiceClient.java [hunk 4]
```diff
@@ -320,54 +328,62 @@ public void createApplicationTemplateFromSP(ServiceProvider serviceProvider, SpT
...
 
         try {
             if (debugEnabled) {
                 log.debug("Deleting Service Provider template: " + templateName);
             }
             stub.deleteApplicationTemplate(templateName);
-        } catch (RemoteException | IdentityApplicationManagementServiceIdentityApplicationManagementException e) {
-            handleException(e);
+        } catch (RemoteException e) {
+            log.error("Error occurred when deleting Service Provider template: " + templateName, e);
```
Comment (DMHP) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/java/org/wso2/carbon/identity/application/mgt/ui/client/ApplicationManagementServiceClient.java]: Shall we add a private method to log and throw the error?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp [hunk 5]
```diff
@@ -387,12 +393,86 @@
             saveTemplate, null);
     }
 
+    function validateTextForIllegal(fld) {
+        var isValid = doValidateInput(fld, "Provided Service Provider Template name is invalid.");
+        if (isValid) {
+            return true;
+        } else {
```
Comment (DMHP) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-service-provider.jsp]: Can we simplify this as below?

 if (isValid) {
            return true;
        }
return false;

**Pull request merged**